flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1846,20 +1861,33 @@ class ScalarExprLowering {
     // procedure argument earlier with no further type information.
     mlir::Value funcPointer;
     mlir::SymbolRefAttr funcSymbolAttr;
+    bool addHostAssociations = false;
     if (const auto *sym = caller.getIfIndirectCallSymbol()) {
       funcPointer = symMap.lookupSymbol(*sym).getAddr();
       assert(funcPointer &&
              "dummy procedure or procedure pointer not in symbol map");
     } else {
       auto funcOpType = caller.getFuncOp().getType();
-      // Deal with argument number mismatch by making a function pointer so that
-      // function type cast can be inserted.
       auto symbolAttr = builder.getSymbolRefAttr(caller.getMangledName());
-      if (callSiteType.getNumResults() != funcOpType.getNumResults() ||
-          callSiteType.getNumInputs() != funcOpType.getNumInputs()) {
-        // Do not emit a warning here because this can happen in legal program
-        // if the function is not defined here and it was first passed as an
-        // argument without any more information.
+      if (callSiteType.getNumResults() == funcOpType.getNumResults() &&
```

Comment:
It would be great to be able to find a funit from a called function symbols if it is defined in the current compilation unit, that would make this kind of handling cleaner. But I know it is not possible right now, so what you have here looks good to me.

Reply:
Yes. As I went through this exercise, there were a couple times I thought it would be convenient to have a backlink to the `funit`.